File: unleash/src/main/java/com/silvercar/unleash/strategy/ConstraintUtil.java
package com.silvercar.unleash.strategy;

import java.util.List;
import com.annimon.stream.Optional;

import com.annimon.stream.Stream;
import com.silvercar.unleash.Constraint;
import com.silvercar.unleash.Operator;
import com.silvercar.unleash.UnleashContext;

public class ConstraintUtil {

    public static boolean validate(List<Constraint> constraints, UnleashContext context) {
        if(constraints != null && constraints.size() > 0) {
            return Stream.of(constraints).allMatch(c -> validateConstraint(c, context));
        } else {
            return true;
        }
    }

    private static boolean validateConstraint(Constraint constraint, UnleashContext context) {
        String contextValue = context.getByName(constraint.getContextName());
        boolean contextValueValid = contextValue != null && !contextValue.isEmpty();
        boolean isIn = contextValueValid && constraint.getValues().contains(contextValue.trim());
        return (constraint.getOperator() == Operator.IN) == isIn;
    }
}

File: unleash/src/main/java/com/silvercar/unleash/strategy/ConstraintUtil.kt
package com.silvercar.unleash.strategy

import com.silvercar.unleash.Constraint
import com.silvercar.unleash.Operator.IN
import com.silvercar.unleash.UnleashContext

class ConstraintUtil {
  fun validate(
    constraints: List<Constraint>?,
    context: UnleashContext
  ): Boolean {
    if (constraints.isNullOrEmpty()) {
      return true
    }
    return constraints.all { validateConstraint(it, context) }
  }

  private fun validateConstraint(
    constraint: Constraint,
    context: UnleashContext
  ): Boolean {
    val contextValue: String = context.getByName(constraint.contextName).trim()
    return constraint.operator == IN == constraint.values.contains(contextValue)
  }
}
